react/features/local-recording/recording/WavAdapter.js

```diff
@@ -0,0 +1,284 @@
+        view.setUint32(4, 10, true);
+
+        // DUMMY data chunk length (set real value on export)
+        view.setUint32(40, 10, true);
+
+        return new Uint8Array(buffer);
+    }
+
+
+    /**
+     * Callback function that saves the PCM bits.
+     *
+     * @private
+     * @param {Float32Array} data - The audio PCM data.
+     * @returns {void}
+     */
+    _saveWavPCM(data) {
+        // need to copy the Float32Array,
+        // unlike passing to WebWorker,
+        // this data is passed by reference,
+        // so we need to copy it, otherwise the
+        // audio file will be just repeating the last
+        // segment.
+        this._wavBuffers.push(new Float32Array(data));
+        this._wavLength += data.length;
+    }
+
+    /**
+     * Combines buffers and export to a wav file.
+     *
+     * @private
+     * @param {*} buffers - The stored buffers.
+     * @param {*} length - Total length (in bytes).
+     * @returns {Blob}
+     */
+    _exportMonoWAV(buffers, length) {
+        // buffers: array with
+        //  buffers[0] = header information (with missing length information)
+        //  buffers[1] = Float32Array object (audio data)
+        //  ...
+        //  buffers[n] = Float32Array object (audio data)
+
+        const dataLength = length * 2; // why multiply by 2 here?
+        const buffer = new ArrayBuffer(44 + dataLength);
+        const view = new DataView(buffer);
+
+        // copy WAV header data into the array buffer
+        const header = buffers[0];
+        const len = header.length;
+
+        for (let i = 0; i < len; ++i) {
+            view.setUint8(i, header[i]);
+        }
+
+        // add file length in header
+        view.setUint32(4, 32 + dataLength, true);
+
+        // add data chunk length in header
+        view.setUint32(40, dataLength, true);
+
+        // write audio data
+        floatTo16BitPCM(view, 44, buffers);
+
+        return new Blob([ view ], { type: 'audio/wav' });
+    }
+}
+
+
+/**
+ * Helper function. Writes a UTF string to memory
+ * using big endianness. Required by WAVE headers.
+ *
+ * @param {ArrayBuffer} view - The view to memory.
+ * @param {*} offset - Offset.
+ * @param {*} string - The string to be written.
+ * @returns {void}
+ */
+function writeUTFBytes(view, offset, string) {
+    const lng = string.length;
+
+    // convert to big endianness
+    for (let i = 0; i < lng; ++i) {
+        view.setUint8(offset + i, string.charCodeAt(i));
+    }
+}
+
+/**
+ * Helper function for converting Float32Array to Int16Array.
+ *
+ * @param {*} output - The output buffer.
+ * @param {*} offset - The offset in output buffer to write from.
+ * @param {*} inputBuffers - The input buffers.
+ * @returns {void}
+ */
+function floatTo16BitPCM(output, offset, inputBuffers) {
+
+    let i, input, isize, s;
+    const jsize = inputBuffers.length;
+    let o = offset;
+
+    // first entry is header information (already used in exportMonoWAV),
```

Comment:
Does the first buffer contain *only* the header?

Reply:
Yes it does.

I find this a little confusing. We have an array of buffers, and the first one is special. It contains the WAV header. The rest contain the PCM in float format. When we export we have to handle the first one in a special way.

Can we keep only PCM in the buffers, and save the header separately (or even better, create it on demand)?

Reply:
Yeah. That's what I am thinking about regarding backing up to disk (the IndexedDB thing).
Basically, currently all the `RecordingAdapter`s have the responsibility of both **recording** and **exporting** the audio. The currently encoded bits ("buffers") are part of the `RecordingAdapter`'s internal state.
With IndexedDB backups, there needs to be a separation between recording/encoding and exporting, because the buffers to be exported will be in IndexedDB rather than `RecordingAdapter`s themselves.